File: app/src/main/java/com/popalay/cardme/ui/addcard/AddCardActivity.java
package com.popalay.cardme.ui.addcard;

import android.app.Dialog;
import android.content.Context;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.view.inputmethod.EditorInfo;
import android.widget.ArrayAdapter;

import com.arellomobile.mvp.presenter.InjectPresenter;
import com.arellomobile.mvp.presenter.ProvidePresenter;
import com.popalay.cardme.R;
import com.popalay.cardme.databinding.ActivityAddCardBinding;
import com.popalay.cardme.ui.base.BaseActivity;
import com.popalay.cardme.utils.DialogFactory;

import java.util.List;

import io.card.payment.CreditCard;

public class AddCardActivity extends BaseActivity implements AddCardView {

    private static final String EXTRA_CARD = "EXTRA_CARD";

    @InjectPresenter AddCardPresenter presenter;

    private ActivityAddCardBinding b;

    public static Intent getIntent(Context context, CreditCard card) {
        final Intent intent = new Intent(context, AddCardActivity.class);
        intent.putExtra(EXTRA_CARD, card);
        return intent;
    }

    @ProvidePresenter
    AddCardPresenter providePresenter() {
        return new AddCardPresenter(getIntent().getParcelableExtra(EXTRA_CARD));
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        b = DataBindingUtil.setContentView(this, R.layout.activity_add_card);
        initUI();
    }

    @Override
    public void showCardDetails(AddCardViewModel vm) {
        b.setModel(vm);
    }

    @Override
    public void setCompletedCardHolders(List<String> holders) {
        final ArrayAdapter<String> adapter = new ArrayAdapter<>(this, android.R.layout.simple_list_item_1, holders);
        b.textHolder.setAdapter(adapter);
    }

    @Override
    public void showError(String message) {
        final Dialog errorDialog = DialogFactory.createCustomButtonsDialog(this, message,
                getString(R.string.action_close), null,
                (dialog, which) -> presenter.onCloseClick(), dialog -> presenter.onErrorDialogDismiss());

        errorDialog.setCancelable(false);
        errorDialog.show();
    }

    private void initUI() {
        setSupportActionBar(b.toolbar);
        b.toolbar.setNavigationOnClickListener(v -> finish());

        b.buttonSave.setOnClickListener(v -> presenter.onAcceptClick(b.getModel().card));

        b.textTitle.setOnEditorActionListener((v, actionId, event) -> {
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                if (b.buttonSave.isEnabled()) {
                    b.buttonSave.performClick();
                }
                return true;
            }
            return false;
        });

    }
}

File: app/src/main/java/com/popalay/cardme/presentation/screens/addcard/AddCardActivity.kt
package com.popalay.cardme.presentation.screens.addcard

import android.arch.lifecycle.ViewModelProvider
import android.arch.lifecycle.ViewModelProviders
import android.content.Context
import android.content.Intent
import android.databinding.DataBindingUtil
import android.os.Bundle
import com.popalay.cardme.R
import com.popalay.cardme.databinding.ActivityAddCardBinding
import com.popalay.cardme.presentation.base.BaseActivity
import com.popalay.cardme.presentation.base.navigation.CustomNavigator
import com.popalay.cardme.utils.DialogFactory
import dagger.android.AndroidInjection
import io.card.payment.CreditCard
import io.reactivex.rxkotlin.addTo
import ru.terrakok.cicerone.NavigatorHolder
import javax.inject.Inject

class AddCardActivity : BaseActivity() {

    @Inject lateinit var navigationHolder: NavigatorHolder
    @Inject lateinit var factory: ViewModelProvider.Factory

    private lateinit var b: ActivityAddCardBinding
    private lateinit var viewModelFacade: AddCardViewModelFacade

    companion object {
        val KEY_CREDIT_CARD = "KEY_CREDIT_CARD"

        fun getIntent(context: Context, creditCard: CreditCard): Intent {
            val intent = Intent(context, AddCardActivity::class.java)
            intent.putExtra(KEY_CREDIT_CARD, creditCard)
            return intent
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        AndroidInjection.inject(this)
        super.onCreate(savedInstanceState)
        b = DataBindingUtil.setContentView<ActivityAddCardBinding>(this, R.layout.activity_add_card)
        ViewModelProviders.of(this, factory).get(AddCardViewModel::class.java).let {
            b.vm = it
            viewModelFacade = it
        }
        initUI()
    }

    private fun initUI() {
        setSupportActionBar(b.toolbar)

        //TODO
        viewModelFacade.doOnShowCardExistsDialog {
            DialogFactory.createCustomButtonsDialog(this@AddCardActivity,
                    R.string.error_card_exist, R.string.action_close,
                    onDismiss = viewModelFacade::onShowCardExistsDialogDismiss)
                    .apply { setCancelable(false) }
        }.addTo(disposables)
    }
}
